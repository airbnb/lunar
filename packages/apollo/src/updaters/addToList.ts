import { DocumentNode } from 'graphql';
import get from 'lodash/get';
import set from 'lodash/set';
// eslint-disable-next-line import/no-extraneous-dependencies
import { DataProxy } from 'apollo-cache';
import { MutationFetchResult } from 'react-apollo';
import prepareQuery from '../utils/prepareQuery';
import getQueryName from '../utils/getQueryName';

export default function addToList<Result, Vars = {}>(
  docOrQuery: DocumentNode | DataProxy.Query<Vars>,
  listPath: string,
  mutationPath: string,
) {
  const query = prepareQuery<Vars>(docOrQuery);

  return (cache: DataProxy, mutationResult: MutationFetchResult<Result>) => {
    const queryResult = cache.readQuery<Result>(query);
    const nextResult = { ...queryResult };
    const list = get(queryResult, listPath);

    if (typeof list === 'undefined' || !Array.isArray(list)) {
      if (__DEV__) {
        throw new TypeError(`"${getQueryName(query.query)}" list "${listPath}" is not an array.`);
      } else {
        return;
      }
    }

    const data = get(mutationResult.data, mutationPath);

    if (typeof data === 'undefined') {
      if (__DEV__) {
        throw new Error(`Cannot find mutation "${mutationPath}". Unable to update query list.`);
      } else {
        return;
      }
    }

    set(nextResult, listPath, [...list, data]);

    cache.writeQuery({
      ...query,
      data: nextResult,
    });
  };
}
