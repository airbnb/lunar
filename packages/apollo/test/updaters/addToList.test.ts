import gql from 'graphql-tag';
import { InMemoryCache } from 'apollo-cache-inmemory';
import addToList from '../../src/updaters/addToList';

const QUERY = gql`
  query getSomething {
    something {
      id
      name
      things {
        id
        type
      }
    }
  }
`;

describe('addToList()', () => {
  let cache: InMemoryCache;

  beforeEach(() => {
    cache = new InMemoryCache();

    cache.writeQuery({
      query: QUERY,
      variables: { id: 123 },
      data: {
        something: {
          id: 123,
          name: 'Something',
          things: [
            { id: 1, type: 'A', __typename: 'thing' },
            { id: 2, type: 'B', __typename: 'thing' },
          ],
          __typename: 'something',
        },
      },
    });
  });

  it('errors if query does not exist in cache', () => {
    expect(() => {
      addToList(
        gql`
          query otherQuery {
            things {
              id
            }
          }
        `,
        'something.things',
        'updateThing',
      )(cache, {});
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if property name is not an array', () => {
    expect(() => {
      // Should be `something.things`
      addToList(QUERY, 'something.name', 'updateThing')(cache, {});
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if mutation data does not exist', () => {
    expect(() => {
      addToList(QUERY, 'something.things', 'updateThing')(cache, {});
    }).toThrowErrorMatchingSnapshot();
  });

  it('adds mutation data to the querys list', () => {
    addToList(
      QUERY,
      'something.things',
      'updateThing',
    )(cache, {
      data: {
        updateThing: {
          id: 3,
          type: 'C',
          __typename: 'thing',
        },
      },
    });

    expect(cache.readQuery({ query: QUERY })).toEqual({
      something: {
        id: 123,
        name: 'Something',
        things: [
          { id: 1, type: 'A', __typename: 'thing' },
          { id: 2, type: 'B', __typename: 'thing' },
          { id: 3, type: 'C', __typename: 'thing' },
        ],
        __typename: 'something',
      },
    });
  });
});
