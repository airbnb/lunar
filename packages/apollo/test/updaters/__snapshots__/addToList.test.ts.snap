// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addToList() errors if mutation data does not exist 1`] = `"Cannot find mutation \\"updateThing\\". Unable to update query list."`;

exports[`addToList() errors if property name is not an array 1`] = `"\\"getSomething\\" list \\"something.name\\" is not an array."`;

exports[`addToList() errors if query does not exist in cache 1`] = `
"Can't find field things on object {
  \\"something\\": {
    \\"type\\": \\"id\\",
    \\"generated\\": false,
    \\"id\\": \\"something:123\\",
    \\"typename\\": \\"something\\"
  }
}."
`;
